gateway/src/metrics.rs

```diff
@@ -0,0 +1,1177 @@
+                &log,
+                "SP sensor metrics configured";
+                "poll_interval" => ?cfg.sp_poll_interval,
+                "max_buffered_sample_chunks" => cfg.max_buffered_sample_chunks,
+            );
+
+            tokio::spawn(
+                PollerManager {
+                    sample_tx,
+                    apictx,
+                    poll_interval: cfg.sp_poll_interval,
+                    tasks: tokio::task::JoinSet::new(),
+                    log,
+                    mgs_id: id,
+                }
+                .run(rack_id_rx),
+            )
+        };
+
+        let (addrs_tx, addrs_rx) =
+            tokio::sync::watch::channel(addresses.clone());
+        let server = {
+            let log = log.new(slog::o!("component" => "producer-server"));
+            let registry = ProducerRegistry::with_id(id);
+            registry
+                .register_producer(Producer { sample_rx, log: log.clone() })
+                .context("failed to register metrics producer")?;
+
+            tokio::spawn(
+                ServerManager { log, addrs: addrs_rx, registry }.run(cfg),
+            )
+        };
+        Ok(Self { addrs_tx, rack_id_tx, server, pollers })
+    }
+
+    pub fn set_rack_id(&mut self, rack_id: Uuid) {
+        if let Some(tx) = self.rack_id_tx.take() {
+            tx.send(rack_id).expect("why has the sensor-poller task gone away?")
+        }
+        // ignoring duplicate attempt to set the rack ID...
+    }
+
+    pub async fn update_server_addrs(&self, new_addrs: &[SocketAddrV6]) {
+        self.addrs_tx.send_if_modified(|current_addrs| {
+            if current_addrs.len() == new_addrs.len()
+                // N.B. that we could make this "faster" with a `HashSet`,
+                // but...the size of this Vec of addresses is probably going to
+                // two or three items, max, so the linear scan actually probably
+                // outperforms it...
+                && current_addrs.iter().all(|addr| new_addrs.contains(addr))
+            {
+                return false;
+            }
+
+            // Reuse existing `Vec` capacity if possible.This is almost
+            // certainly not performance-critical, but it makes me feel happy.
+            current_addrs.clear();
+            current_addrs.extend_from_slice(new_addrs);
+            true
+        });
+    }
+}
+
+impl Drop for Metrics {
+    fn drop(&mut self) {
+        // Clean up our children on drop.
+        self.server.abort();
+        self.pollers.abort();
+    }
+}
+
+impl MetricsConfig {
+    fn validate(self) -> anyhow::Result<ValidatedMetricsConfig> {
+        anyhow::ensure!(
+            self.oximeter_collection_interval_secs > 0,
+            "`metrics.oximeter_collection_interval_secs` probably shouldn't \
+             be 0 seconds",
+        );
+        let oximeter_collection_interval =
+            Duration::from_secs(self.oximeter_collection_interval_secs as u64);
```

Comment:
Similar to my question above, why use a `usize` in the config if we're casting to `u64` here?